spec/client/ics-007-tendermint-client/README.md

```diff
@@ -184,69 +196,133 @@ function latestClientHeight(clientState: ClientState): Height {
-    clientState.latestHeight = header.height
-    // update latest timestamp
-    clientState.latestTimestamp = header.timestamp
+    assert(verify(
+      header.trustedValidatorSet,
+      clientState.latestHeight,
+      clientState.trustingPeriod,
+      clientState.maxClockDrift,
+      header.TendermintSignedHeader,
+    ))
+}
+```
+
+### CheckForMisbehaviour
+
+CheckForMisbehaviour will check if an update contains evidence of Misbehaviour. If the ClientMessage is a header we check for implicit evidence of misbehaviour by checking if there already exists a conflicting consensus state in the store or if the header breaks time monotonicity.
+
+```typescript
+function checkForMisbehaviour(
+  clientState: clientState,
+  clientMsg: clientMessage) => bool {
+    switch typeof(clientMsg) {
+      case Header:
+        // fetch consensusstate at header height if it exists
+        consensusState = get("clients/{identifier}/consensusStates/{header.Height}")
+        // if consensus state exists and conflicts with the header
+        // then the header is evidence of misbehaviour
+        if consensusState != nil && consensusState.commitmentRoot != header.AppHash {
+          return true
+        }
+
+        // check for time monotonicity misbehaviour
+        // if header is not monotonically increasing with respect to neighboring consensus states
+        // then return true
+        // NOTE: implementation must have ability to iterate ascending/descending by height
+        prevConsState = getPreviousConsensusState(header.Height)
+        nextConsState = getNextConsensusState(header.Height)
+        if prevConsState.timestamp >= header.timestamp {
+          return true
+        }
+        if nextConsState != nil && nextConsState.timestamp <= header.timestamp {
+          return true
+        }
+      case Misbehaviour:
+        // assert that the heights are the same
+        assert(misbehaviour.h1.height === misbehaviour.h2.height)
+        // assert that the commitments are different
+        assert(misbehaviour.h1.commitmentRoot !== misbehaviour.h2.commitmentRoot)
+    }
+}
+```
+
+### UpdateState
+
+UpdateState will perform a regular update for the Tendermint client. It will add a consensus state to the client store. If the header is higher than the lastest height on the clientState, then the clientState will be updated.
+
+```typescript
+function updateState(
+  clientState: clientState,
+  clientMsg: clientMessage) {
+    header = Header(clientMessage)
+    // only update the clientstate if the header height is higher
+    // than clientState latest height
+    if clientState.height < header.height {
+      // update validator set
+      clientState.validatorSet = header.validatorSet
+      // update latest height
+      clientState.latestHeight = header.height
+      // update latest timestamp
+      clientState.latestTimestamp = header.timestamp
+
+      // save the client
+      set("clients/{identifier}", clientState)
```

Comment:
In [ICS-24](https://github.com/cosmos/ibc/blob/main/spec/core/ics-024-host-requirements/README.md#path-space), the key for `ClientState`s is defined as `clients/{identifier}/clientState`